src/primitive_shape.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 the Kurbo Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#![allow(unused_qualifications)]
+
+use crate::{Point, Rect, Shape};
+use alloc::boxed::Box;
+
+mod _never_shape {
+    use super::*;
+    use crate::PathEl;
+    /// An uninhabited type that implements shape.
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    #[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+    #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+    pub enum NeverShape {}
+    impl Shape for NeverShape {
+        type PathElementsIter<'a> = core::iter::Empty<PathEl>;
+        fn path_elements(&self, _: f64) -> Self::PathElementsIter<'_> {
+            unreachable!()
+        }
+        fn area(&self) -> f64 {
+            unreachable!()
+        }
+        fn perimeter(&self, _: f64) -> f64 {
+            unreachable!()
+        }
+        fn winding(&self, _: Point) -> i32 {
+            unreachable!()
+        }
+        fn bounding_box(&self) -> Rect {
+            unreachable!()
+        }
+    }
+}
```

Comment:
I would recommend removing this block and the whole `External` shape argument.

The External argument only lets you add one other shape to `ConcreteShape` anyway. If you want to add multiple "extra" shapes, you need to pass an enum as the `External` argument, at which point you might as well just implement you own enum instead of ConcreteShape.

Reply:
My general  feeling is that by using the defaulted type parameter it really doesn't add too much in the way of extra complexity the caller needs to be aware of (unfortunately it's still there to be looked at).

But does preserve the openness of type, meaning it'll becomes possible to use in library code even though someone is using an external shape argument, or across projects which both include extra shapes by making an enum which bridges both.

The need for an extra enum is not desirable, but it seems the only way to make an open enum type.
Without this I fear that someone will want to use the code in library code, and rather than *might as well* they *must*  implement their own enum instead of using ConcreteShape. Then once people start targeting their own enum as the base of their API it becomes hard to share code across projects that depend upon kurbo.

So I would kind of push back, but I do think it is useful without the extra type parameter, and that custom shape impls seem pretty rare, so I would probably relent if there are strong feelings against this type parameter.

Reply:
Fwiw, I also strongly agree with Olivier here. Then enums that these people will make would likely be `enum {StaticShape, MyShape}`; the scenario where that's an issue is exactly the same scenario where your propsed `StaticShape<MyShape>` becomes an issue.

(this is apart from the merit of this API at all, on which I have no opinion)

Reply:
Sounds like I'm in the minority here then, and I value consensus more than my own opinion, So I'll probably give it some time to see if anyone else chimes in, then remove it.

By removing it I guess what we lose is the ability to `StaticShape<T> where T: Shape` from library code, which is kind of what I was aiming for, along with avoiding a proliferation of shape wrappers.